### ISSUE TITLE
Proposal: no-useless-else rule

### ISSUE THREAD

**Issue by Trott** [eslint/eslint#8028]
**Please describe what the rule should do:**

If an `if` block contains a `throw` statement, the `else` block becomes unnecessary. Its contents can be placed outside of the block.

**What category of rule is this? (place an "X" next to just one item)**

[x] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// With this rule, this would cause ESLint to warn:

if (foo)
    throw new Error('foo');
else
    return bar;

// This rule would *not* warn about this:

if (foo)
    throw new Error('foo');

return bar;
```

**Why should this rule be included in ESLint (instead of a plugin)?**

It seems generally applicable. It is analogous to the currently-include `no-else-return` rule.

**Comment by not-an-aardvark** [comment]
:+1:, but I wonder if it would be better to just expand the `no-else-return` rule to include all statements that cause control flow jumps (`return, `throw`, `break` and `continue`).

Maybe we could rename it to `no-useless-else` or something.

**Comment by kaicataldo** [comment]
I'm in agreement with @not-an-aardvark.

**Comment by mysticatea** [comment]
I like `no-useless-else` idea.

It might be possible to remove `else` clauses if the end of previous `if` clause is unreachable in code path analysis.

**Comment by platinumazure** [comment]
Which proposal should we be evaluating? `no-else-throw` or `no-useless-else`?

@Trott Would you be okay with us implementing `no-useless-else` per @not-an-aardvark's proposal above, which should take care of your use case and others besides?

**Comment by Trott** [comment]
> @Trott Would you be okay with us implementing no-useless-else per @not-an-aardvark's proposal above, which should take care of your use case and others besides?

That works. Even better if there is a configuration option to turn on and off detection for each type of flow control statement (`ThrowStatement`, `ReturnStatement`, etc.).

**Comment by not-an-aardvark** [comment]
@eslint/eslint-team The proposal from https://github.com/eslint/eslint/issues/8028#issuecomment-277498571 has 6 :+1:s. Is anyone willing to champion it? If not, we should close this issue.

**Comment by platinumazure** [comment]
I'll champion.

I can probably write a PR for this either this weekend or next weekend, but if anyone wants to jump on this now, please feel free! (Just please leave a comment like "Working on this" so I know, thanks!)

**Comment by not-an-aardvark** [comment]
@platinumazure If you implement this, feel free to use some of my old WIP commits from [here](https://github.com/not-an-aardvark/eslint/tree/no-useless-else-wip). I was originally working on this a bit back in February. The first commit involves renaming the rule and adding some tests, and the second commit is from an old `git stash` entry where I tried to make the rule use code path analysis to catch some extra cases, but didn't end up finishing.

(Disclaimer: Aside from those details, I don't remember what state those commits were in -- it might not have been PR-ready.)

**Comment by MatthewHerbst** [comment]
@platinumazure / @not-an-aardvark Any status update on that WIP? I'd be happy to try to help out if needed

**Comment by ilyavolodin** [comment]
@MatthewHerbst I think if you want to take over this issue, both @not-an-aardvark and @platinumazure would appreciate it, since there hasn't been any movement on this in a while.

**Comment by platinumazure** [comment]
Are we anywhere near accepting this proposal? Seems to me we should close
this, unless we can get the team to support it right away.

**Comment by ilyavolodin** [comment]
It looks like there's enough support for this issue, just needs a champion to be accepted.

**Comment by platinumazure** [comment]
@ilyavolodin I'm championing-- we need one more :+1:.

Although: I just want to confirm, we're going with @not-an-aardvark's proposal, right? Should we edit the original issue post to reflect this and/or create a new issue for that proposal, to avoid confusion?

**Comment by btmills** [comment]
@platinumazure I added one more :+1: so we can accept this, and yes, going with @not-an-aardvark's `no-useless-else` proposal. @Trott would you be willing to update your original issue to use `no-useless-else`?

**Comment by Trott** [comment]
>  @Trott would you be willing to update your original issue to use no-useless-else?

@btmills I think the only place it needs updating is the title of this issue, so I updated that. If there's somewhere else it needs updating, I'm happy to do it (or, since you have Owner role, please feel free to do it for me).

**Comment by platinumazure** [comment]
This proposal is accepted but has sat around for a while. I'm going to add the "help wanted" label to try to draw some attention from the community.

@Trott I think we've updated everything correctly here. Would you be interested in writing a PR? If not, no worries, just asking.

**Comment by Trott** [comment]
@platinumazure I could try. I imagine it would involve:

* reproducing all the `no-else-return` tests for the `no-useless-else` rule
* adding tests for `throw` rather than just `return` in the `if` block
* optionally adding tests for `break` and `continue` as well if applicable
* implement `lib/rules/no-else-return.js` such that it passes the tests
* write documentation for `no-else-return`

What about the `no-else-return` rule? Does it need to be deprecated somehow as part of this? Or is that a separate thing? We'd have to be careful to make sure that rules can't be configured to contradict each other, of course, but that's standard procedure for ESLint I'm sure.

**Comment by Mouvedia** [comment]
```
if (foo)
    throw new Error('foo');
else if (qux) // should be if (qux)
    return bar;
return baz;
```
should warn as well.

**Comment by AMilassin** [comment]
What about this special case? I think no-else-return should support this too, but if no-else-return learns about `throw`, does the need for making a distinction between no-useless-else and no-else-return vanishes?

This is currently a no-go with no-else-return, it reports it as an error even though adhering to the rule results in an incorrect/invalid code.
```
if (foo) {
   return foo;
} else if (bar) (
   return bar;
}
throw new Error('Unsupported case');
```

**Comment by Trott** [comment]
I don't think I'm going to get around to doing this any time too soon. Feel free to re-open if someone else is likely to pick it up.
